src/MemoryTracing/MemoryTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,145 @@
+    listener_.reset();
+    return events;
+  }
+
+ private:
+  uint64_t memory_sampling_period_ns_;
+  std::optional<MemoryInfoProducer> producer_ = std::nullopt;
+  std::optional<BufferMemoryInfoListener> listener_ = std::nullopt;
+};
+
+[[nodiscard]] std::vector<ProducerCaptureEvent> TraceAndGetEvents(
+    MemoryTracingIntegrationTestFixture* fixture) {
+  CHECK(fixture != nullptr);
+
+  fixture->StartTracingAndWaitForTracingLoopStarted();
+
+  constexpr absl::Duration kSleepBeforeStopTracing = absl::Milliseconds(1000);
+  absl::SleepFor(kSleepBeforeStopTracing);
+  return fixture->StopTracingAndGetEvents();
+}
+
+void VerifyOrderAndContentOfEvents(const std::vector<ProducerCaptureEvent>& events,
+                                   uint64_t sampling_period_ns) {
+  const uint64_t kSamplingTimeErrorToleranceNs = absl::Milliseconds(1) / absl::Nanoseconds(1);
+  uint64_t previous_event_timestamp_ns = 0;
+  for (const auto& event : events) {
+    if (event.event_case() != ProducerCaptureEvent::kSystemMemoryUsage) UNREACHABLE();
+
+    // Verify whether events are in order of their timestamps.
+    uint64_t current_event_timestamp_ns = event.system_memory_usage().timestamp_ns();
+    EXPECT_GE(current_event_timestamp_ns, previous_event_timestamp_ns);
+
+    // Verify whether the sampling period works as expected.
+    if (previous_event_timestamp_ns > 0 && sampling_period_ns > 0) {
+      uint64_t time_since_last_successful_sampling =
+          current_event_timestamp_ns - previous_event_timestamp_ns;
+      uint64_t periods_since_last_successful_sampling =
+          time_since_last_successful_sampling / sampling_period_ns;
+      // If there are failed attempts of memory sampling, the sampling time error would be
+      // difficult to decide. Hence, we only check the case of successive successful attempts of
+      // sampling.
+      if (periods_since_last_successful_sampling <= 1) {
+        EXPECT_GE(time_since_last_successful_sampling,
+                  sampling_period_ns - kSamplingTimeErrorToleranceNs);
+        EXPECT_LE(time_since_last_successful_sampling,
+                  sampling_period_ns + kSamplingTimeErrorToleranceNs);
+      }
+    }
+
+    // Verify whether the contents of events are valid.
+    EXPECT_TRUE(event.system_memory_usage().total_kb() == kMissingInfo ||
+                event.system_memory_usage().total_kb() > 0);
+    EXPECT_TRUE(event.system_memory_usage().free_kb() == kMissingInfo ||
+                event.system_memory_usage().free_kb() > 0);
+    EXPECT_TRUE(event.system_memory_usage().available_kb() == kMissingInfo ||
+                event.system_memory_usage().available_kb() > 0);
+    EXPECT_TRUE(event.system_memory_usage().buffers_kb() == kMissingInfo ||
+                event.system_memory_usage().buffers_kb() > 0);
+    EXPECT_TRUE(event.system_memory_usage().cached_kb() == kMissingInfo ||
+                event.system_memory_usage().cached_kb() > 0);
```

Comment:
Do we actually normally expect information to be missing? I would expect them to always be there and hence the part after the `||` to be enough.

Reply:
Thanks for raising this issue.

In the `ParseMemInfo` method, I initialize the fields in the `SystemMemoryUsage` event with `kMissingInfo`. If we assume the `proc/meminfo` file is always well-formatted and so we can successfully extract the information we need, should I change the logic in the `ParseMemInfo` method? I'm thinking that the expected information in the test and in the method where we produce the event should be the same.

Reply:
Thanks for the explanation! I removed the comparison with `kMissingInfo`.